libc/string_mem.c

```diff
@@ -0,0 +1,396 @@
+	TEST_ASSERT_EQUAL_MEMORY(TEST_STR, ptr, TEST_STR_SIZE);
+	TEST_ASSERT_EQUAL_MEMORY(TEST_STR + OVERLAP_LEN, ptr + TEST_STR_SIZE, OVERLAP_LEN);
+	TEST_ASSERT_EACH_EQUAL_INT8(0, mem, OFFSET - OVERLAP_LEN);
+	TEST_ASSERT_EACH_EQUAL_INT8(0, mem + OFFSET + TEST_STR_SIZE, MEM_SIZE - (OFFSET + TEST_STR_SIZE));
+}
+
+
+TEST(string_memmove, move_nothing)
+{
+	uint8_t *ptr = mem + OFFSET;
+	/* Check if any data is moved */
+	ret = memmove(ptr, TEST_STR, 0);
+	TEST_ASSERT_EQUAL_PTR(ptr, ret);
+	TEST_ASSERT_EQUAL_MEMORY(TEST_STR, ptr, TEST_STR_SIZE);
+	TEST_ASSERT_EACH_EQUAL_INT8(0, mem, OFFSET);
+	TEST_ASSERT_EACH_EQUAL_INT8(0, mem + OFFSET + TEST_STR_SIZE, MEM_SIZE - (OFFSET + TEST_STR_SIZE));
+}
+
+
+TEST(string_memmove, move_byte)
+{
+#if __BYTE_ORDER == __LITTLE_ENDIAN
+	uint32_t buffer = 0x0000FF00;
+	ret = memmove(&buffer, (uint8_t *)&buffer + 1, 1);
+	TEST_ASSERT_EQUAL_PTR(&buffer, ret);
+	TEST_ASSERT_EQUAL_INT32(0x0000FFFF, buffer);
+
+	buffer = 0x0000FF00;
+	ret = memmove((int8_t *)(&buffer) + 3, (uint8_t *)&buffer + 1, 1);
+	TEST_ASSERT_EQUAL_PTR((int8_t *)(&buffer) + 3, ret);
+	TEST_ASSERT_EQUAL_INT32(0xFF00FF00, buffer);
+
+	buffer = 0x0000FF00;
+	ret = memmove(((int8_t *)(&buffer)) + 1, (uint8_t *)&buffer + 1, 1);
+	TEST_ASSERT_EQUAL_PTR(((int8_t *)(&buffer)) + 1, ret);
+	TEST_ASSERT_EQUAL_INT32(0x0000FF00, buffer);
+#else
+	uint32_t buffer = 0x00FF0000;
+	ret = memmove(&buffer, (uint8_t *)&buffer + 1, 1);
+	TEST_ASSERT_EQUAL_PTR(&buffer, ret);
+	TEST_ASSERT_EQUAL_INT32(0xFFFF0000, buffer);
+
+	buffer = 0x0000FF00;
+	ret = memmove((int8_t *)(&buffer) + 3, (uint8_t *)&buffer + 1, 1);
+	TEST_ASSERT_EQUAL_PTR((int8_t *)(&buffer) + 3, ret);
+	TEST_ASSERT_EQUAL_INT32(0x00FF00FF, buffer);
+
+	buffer = 0x0000FF00;
+	ret = memmove(((int8_t *)(&buffer)) + 1, (uint8_t *)&buffer + 1, 1);
+	TEST_ASSERT_EQUAL_PTR(((int8_t *)(&buffer)) + 1, ret);
+	TEST_ASSERT_EQUAL_INT32(0x0000FF00, buffer);
+#endif
+}
+
+
+TEST_SETUP(string_memmove_big)
+{
+	mem_big = (uint8_t *)calloc(MEM_SIZE_BIG, 1);
+	TEST_ASSERT_NOT_NULL(mem_big);
+	ret = memcpy(mem_big + OFFSET_BIG, big_buf, BIG_SIZE);
+	TEST_ASSERT_EQUAL_PTR(mem_big + OFFSET_BIG, ret);
+	TEST_ASSERT_EACH_EQUAL_INT8(0, mem_big, OFFSET_BIG);
+	TEST_ASSERT_EACH_EQUAL_INT8(0, mem_big + OFFSET_BIG + BIG_SIZE, MEM_SIZE_BIG - (OFFSET_BIG + BIG_SIZE));
+}
+
+
+TEST_TEAR_DOWN(string_memmove_big)
+{
+	free(mem_big);
+}
+
+
+TEST(string_memmove_big, no_overlap_end)
+{
+	uint8_t *ptr = mem_big + (MEM_SIZE_BIG - BIG_SIZE);
+	/* Move to the end of allocated memory */
+	ret = memmove(ptr, mem_big + OFFSET_BIG, BIG_SIZE);
+	TEST_ASSERT_EQUAL_PTR(ptr, ret);
+	TEST_ASSERT_EQUAL_MEMORY(big_buf, ptr, BIG_SIZE);
+	TEST_ASSERT_EQUAL_MEMORY(big_buf, mem_big + OFFSET_BIG, BIG_SIZE);
+	TEST_ASSERT_EACH_EQUAL_INT8(0, mem_big + OFFSET_BIG + BIG_SIZE, (MEM_SIZE_BIG - BIG_SIZE) - (OFFSET_BIG + BIG_SIZE));
+	TEST_ASSERT_EACH_EQUAL_INT8(0, mem_big, OFFSET_BIG);
+}
+
+
+TEST(string_memmove_big, no_overlap_between)
+{
+	int move_offset = 2824;
```

Comment:
`const`

I don't see the relation to OFFSET_BIG and BIG_SIZE, maybe write the value using defined variables?

Reply:
This is arbitrary value so that the block is not adjacent to TEST_STR and end of buffer.
As below:

|...............|TEST_STR|.........................|MOVED_STR|..................|
0--------1500----------2524---------------2824------------3848---------4000

Reply:
I know I can analyze it, the comment was regarding the code readability and possible issues when somebody would change the define values in the future
```c
const uint8_t *src = mem_big + OFFSET_BIG;
const uint32_t gapSize = 100;

/* ensure we have enough place in the buffer */
if (src + BIG_SIZE + gapSize + BIG_SIZE > mem_big + MEM_SIZE_BIG) {
  FAIL("Invalid test setup");
}
memcpy(src + BIG_SIZE + gapSize, src, BIG_SIZE);  // <--- clearly readable that I will have 2 copies of data separated by gapSize - without extra comments / ascii drawings (but i would appreciate them nevertheless ;))
/* ... */